src/com/google/javascript/jscomp/AbstractCommandLineRunner.java

```diff
@@ -1130,35 +1267,58 @@ private void outputModuleBinaryAndSourceMaps(
     // If the source map path is in fact a pattern for each
     // module, create a stream per-module. Otherwise, create
     // a single source map.
-    Writer mapOut = null;
+    Writer mapFileOut = null;
 
-    if (!shouldGenerateMapPerModule(options)) {
-      mapOut = fileNameToOutputWriter2(expandSourceMapPath(options, null));
+    // When the json_streams flag is specified, sourcemaps are always generated
+    // per module
+    if (!(shouldGenerateMapPerModule(options) || config.useJsonStreams)) {
+      mapFileOut = fileNameToOutputWriter2(
+          expandSourceMapPath(options, null));
     }
 
     for (JSModule m : modules) {
-      if (shouldGenerateMapPerModule(options)) {
-        mapOut = fileNameToOutputWriter2(expandSourceMapPath(options, m));
-      }
+      if (config.useJsonStreams) {
```

Comment:
here, the result of the IF is the same in all iterations of the loop. So the IF can move outside the loop. And the json-stream stuff can be extracted into a helper method to make this easier to read.

    if (config.useJsonStreams) {
      helper-method;
    } else {
      for (JSModule m : modules) {
        previous-stuff;
      }
    }